### PR TITLE
Making direct-C and direct-Python agree

### PR DESCRIPTION
**Pull request by DanHickstein**
This addresses https://github.com/PyAbel/PyAbel/issues/58. This is still a work in progress, but I'm starting a PR to facilitate comparison of the changes.

### CODE REVIEW

File: abel/direct.py [hunk 1]
```diff
@@ -187,17 +188,24 @@ def _pyabel_direct_integral(f, r, correction, int_func=simpson_rule_wrong):
...
+    mask = (II < JJ) 
 
     I_sqrt = np.zeros(R.shape)
     I_sqrt[mask] = np.sqrt((Y**2 - R**2)[mask])
 
     I_isqrt = np.zeros(R.shape)
     I_isqrt[mask] = 1./I_sqrt[mask]
+    
+    mask2 = ((II>JJ-2) & (II < JJ+1))
+    print (mask2)
```
Comment (stggh) [inline abel/direct.py]: delete legacy trace print

File: abel/direct.py [hunk 2]
```diff
@@ -73,7 +74,7 @@ def _construct_r_grid(n, dr=None, r=None):
 
 def direct_transform(fr, dr=None, r=None, direction='inverse',
                      derivative=gradient,
-                     int_func=simpson_rule_wrong,
+                     int_func=np.trapz,
```
Comment (stggh) [inline abel/direct.py]: Perhaps, the change to `int_func` should occur at the source level?
`def _pyabel_direct_integral(f, r, correction, int_func=simpson_rule_wrong):` and delete the redundant `simpson_rule_wrong()`. This code may be placed in the `PyAbelLegacy` repo

**Comment by stggh** [comment]
Wahoo! A good start to 2018 @DanHickstein. Your change to `trapz` appears to work fine. 

Should the legacy `simpson_rule_wrong`and references be deleted?

Also, update the inline docs

**Comment by stggh** [comment]
travis-ci build failing due to a disk space issue. Note for #201 I switched to `pip wheel` to bulid 2.7. However. the disk space error may be unrelated: `.travis.yml`:
```
install:
- pip install --upgrade pip setuptools wheel
- pip install --only-binary=numpy,scipy numpy scipy
- pip install coveralls
- python setup.py install
script:
```

File: abel/direct.py (hunk 2 above)
Comment (DanHickstein) [reply abel/direct.py]: I think that it's fine to delete the simpson_rule_wrong function. `scipy.integrate.simps` can also handle both even and odd sizes, so I'm unsure exactly why this function is needed.

File: abel/direct.py [hunk 3]
```diff
@@ -107,6 +89,13 @@ def direct_transform(fr, dr=None, r=None, direction='inverse',
         if True in addition to the integration with the Simpson rule,
         integration near the singular value is done analytically,
         assuming a piecewise linear data.
+    backend : string
```
Comment (stggh) [inline abel/direct.py]: missed the `f` -> `r` on line 77:
>f : 1D ndarray
        the spatial mesh (optional)

Comment (DanHickstein) [reply abel/direct.py]: ah, okay, got it now.

**Comment by DanHickstein** [comment]
I think you are correct about `reflect_array()`. I'll go ahead and get rid of it.

**Comment by DanHickstein** [comment]
Same thing for `abel_sym` too it seems.

**Comment by stggh** [comment]
Agreed.

**Comment by stggh** [comment]
`tests/test_direct.py` remove `test_simps_wrong()`.

**Comment by DanHickstein** [comment]
Yeah, I was just removing that, and then it sent me spiraling down the hole of trying to get the C/Python agreement test to work. Seems like I'm still off by one pixel on the high-r-value side (which is the side that typically doesn't matter much, but I'd still like to get it to work...)

**Comment by DanHickstein** [comment]
Phew! I emerged from the rabbit hole! 

Direct-C just needed a small tweak to match Direct-Python for the edge case of `r=N`.

I unleashed the previously disabled unit tests. We now have agreement between the Direct-C and Direct-Python methods better than `1e-9`, and the Direct methods reproduce a 501-point gaussian to within 2%.

**Comment by stggh** [comment]
Impressive!  

The build fails due to a missing miniconda package (for 2.7):
```
0.28s$ conda config --set always_yes yes --set changeps1 no
Traceback (most recent call last):
  File "/home/travis/miniconda2/bin/conda", line 11, in <module>
    sys.exit(main())
  File "/home/travis/miniconda2/lib/python2.7/site-packages/conda/cli/main.py", line 125, in main
    from ..exceptions import conda_exception_handler
  File "/home/travis/miniconda2/lib/python2.7/site-packages/conda/exceptions.py", line 17, in <module>
    from .common.io import timeout
  File "/home/travis/miniconda2/lib/python2.7/site-packages/conda/common/io.py", line 5, in <module>
    from concurrent.futures import ThreadPoolExecutor
ImportError: No module named concurrent.futures
```

which may be "fixed" by using the [pip wheel above](https://github.com/PyAbel/PyAbel/pull/204#issuecomment-354680980)  ;-)

**Comment by DanHickstein** [comment]
I'll give that a shot now...

**Comment by DanHickstein** [comment]
Hmmm, when I switch to your code in travis.yml, then it seems that Cython is not installed.

**Comment by stggh** [comment]
One very minor English spelling correction `abel/tests/test_direc.py`  correspondance -> correspondence, although sometimes it is nice to keep the author spelling (@rth) to showcase the multinational nature of `PyAbel`

**Comment by DanHickstein** [comment]
Okay, I made that change. And now I included cython in the pip install, which seems to work.

Now, the Cython extensions are built, but for some reason Travis thinks that the Direct-C method is not available and skips those tests.

File: abel/direct.py (hunk 2 above)
Comment (rth) [reply abel/direct.py]: I agree that it's better to use the trapz/simps functions from scipy if possible.

File: abel/lib/direct.pyx [hunk 4]
```diff
@@ -46,48 +46,38 @@ cpdef _cabel_direct_integral(double [:, ::1] f, double [::1] r, int correction):
...
+        
+                s = 0
+            
+                # simple left Reimann sum:
+                # for k in range(j+1, N1): # inner loop over elements such as r < y
+                #   s = s + f[i,k] * I_isqrt[j,k] * dr
+            
+                # Trapezoidal rule integration, skipping r == y
+                for k in range(j+1, N1-1): # inner loop over elements such as r < y
+                    s = s + f[i,k] * I_isqrt[j,k] * 0.5 * dr  + f[i,k+1] * I_isqrt[j,k+1] * 0.5 * dr
```
Comment (rth) [inline abel/lib/direct.pyx]: maybe use inplace add and factorize multiplications?
```py
s += (f[..] * I_isqrt[..] + f[..] * I_isqrt[..]) * 0.5 * dr
```

File: abel/tests/test_direct.py [hunk 5]
```diff
@@ -8,109 +8,77 @@
...
-    recon = abel.direct.direct_transform(x, direction="inverse")
-
+    recon = abel.direct.direct_transform(x, direction='inverse')
     assert recon.shape == (n, n)
 
 
 def test_direct_zeros():
     # just a sanity check
-    if not abel.direct.cython_ext:
-        raise SkipTest
```
Comment (rth) [inline abel/tests/test_direct.py]: This was intended to skip this test if the cython implemented failed to compile. We don't want this anymore?

File: abel/direct.py [hunk 6]
```diff
@@ -224,45 +222,3 @@ def is_uniform_sampling(r):
...
-    global_assumptions.add(Q.is_true(r2 > y))
-    global_assumptions.add(Q.is_true(r2 > r1))
-    P = c0 + (r-y)*c_r  # + (r-r0)**2*c_rr
-    K_d = 1/sqrt(r**2-y**2)
-    res = integrate(P*K_d, (r, y, r1))
-    sres = simplify(res)
-    print(sres)
-
-
-def reflect_array(x, axis=1, kind='even'):
```
Comment (rth) [inline abel/direct.py]: I think we still have code depending on this function in my former lab, so I would rather keep it if possible to avoid breaking things.

Comment (stggh) [reply abel/direct.py]: Perhaps, the function may be added to `abel/tools/symmetry.py` ?

**Comment by stggh** [comment]
@DanHickstein once addressing @rth's concerns, this PR is ready to merge. Feel free to do so, and I will rebase PR #201, the code changes in which are not really that much.

File: abel/lib/direct.pyx (hunk 4 above)
Comment (stggh) [reply abel/lib/direct.pyx]: Very clever!

File: abel/direct.py (hunk 6 above)
Comment (rth) [reply abel/direct.py]: @stggh Sounds good to me.

File: abel/lib/direct.pyx (hunk 4 above)
Comment (DanHickstein) [reply abel/lib/direct.pyx]: Sounds good to me!

File: abel/direct.py (hunk 6 above)
Comment (DanHickstein) [reply abel/direct.py]: I made the move to `tools.symmetry.reflect_array`. But basically this is just a wrapper for np.flipud and np.fliplr?

If you really don't want to break the code in your previous lab, should we leave the function exactly where it is and include a depreciation warning?

**Comment by DanHickstein** [comment]
Hmmm, the Cython extensions seem to build just fine, but then, when running on Travis, it seems like they cannot be found. I'm not quite sure what's going on. I'll try switching back to the previous Travis.yml

**Comment by DanHickstein** [comment]
Well, I think that the issue on Travis has something to do with where the Cython extensions are put after they are built.

Switching `python setup.py install` to `python setup.py develop` fixed the problem. 

Also, using 
```
python setup.py build_ext -i
python setup.py install
```
works. The first line is to "build extension inplace".

Probably we need to change something about how the extensions are built in the setup.py to ensure that they are properly installed on linux with just `python setup.py install`.

File: abel/direct.py (hunk 6 above)
Comment (rth) [reply abel/direct.py]: OK, I checked, a copy is still available in the [`luli/hedp`](https://github.com/luli/hedp/blob/master/hedp/math/abel.py#L107) repo, so feel free to either remove it or move it under `tools.symmery`, as you prefer. Thanks for fixing this issue, in general!

**Comment by rth** [comment]
> Probably we need to change something about how the extensions are built in the setup.py to ensure that they are properly installed on linux with just python setup.py install

I'll try to have a look once this PR is merged..

File: abel/direct.py (hunk 6 above)
Comment (DanHickstein) [reply abel/direct.py]: Okay, I think that I will remove the function. Thanks!

**Comment by DanHickstein** [comment]
Thanks for the input @rth and @stggh! Are there further suggestions, or should I go ahead and merge later today?

**Comment by stggh** [comment]
Looks good to me.

BTW, on my Linux system `python3 setup.py install --user` builds and finds the cython stuff.
```
nosetests  abel/tests/  --verbosity=2  --with-coverage --cover-package=abel
  :
Ran 40 tests in 3.742s

OK
```

and `direct` `backend='c'` works, so the installation [method issue](https://github.com/PyAbel/PyAbel/pull/204#issuecomment-354845103) may be fine.

**Comment by rth** [comment]
Merged. @DanHickstein thanks again for fixing this and @stggh for the review!

**Comment by DanHickstein** [comment]
Thanks guys!

@stggh, does the install still work if you omit the `--user` bit?

**Comment by stggh** [comment]
Yes

**Comment by DanHickstein** [comment]
Hmm. Okay, maybe that issue is unique to Travis. That's weird. But I suppose it means that we don't need to fix it.
